### PR TITLE
Migrate assets pallet to new macros

### PR DESCRIPTION
**Pull request by dvdplm**
Converts the Assets pallet to the new pallet attribute macro introduced in #6877.

Part of #7882.

[Upgrade guidelines used](https://crates.parity.io/frame_support/attr.pallet.html#upgrade-guidelines).

## :warning: Breaking Change :warning:

From https://crates.parity.io/frame_support/attr.pallet.html#checking-upgrade-guidelines

> storages now use PalletInfo for module_prefix instead of the one given to `decl_storage`: use of this pallet in `construct_runtime!` needs careful updating of the name in order to not break storage or to upgrade storage (moreover for instantiable pallet). If pallet is published, make sure to warn about this breaking change.

So users of the `Assets` pallet must be careful about the name they used in `construct_runtime!`. Hence the `runtime-migration` label, which might not be needed depending on the configuration of the `Assets` pallet.

### CODE REVIEW

**Comment by dvdplm** [comment]
@ascjones Do you know how we should handle the deprecation of `RawEvent`? Should we still provide the type alias or rather try to track down uses and update (or both)?

**Comment by thiolliere** [comment]
> @ascjones Do you know how we should handle the deprecation of `RawEvent`? Should we still provide the type alias or rather try to track down uses and update (or both)?

we can also have a deprecated alias. This could help other to migrate.

(
before this PR raw event was: 
```
enum RawEvent<AccountId, Balance, AssetId> {...};
type Event<C: Config> = RawEvent<C::AccounId,...>;
```
)
Maybe we can now do:
```
#[deprecated(...)]
type RawEvent<T> = Event<T>;
```
But at the same time this is a breaking RawEvent as its generic are different

File: frame/assets/src/lib.rs [hunk 1]
```diff
@@ -114,214 +114,81 @@ mod benchmarking;
...
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(PhantomData<T>);
 
-	/// The currency mechanism.
-	type Currency: ReservableCurrency<Self::AccountId>;
+	#[pallet::config]
+	/// The module configuration trait.
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + Into<<Self as frame_system::Config>::Event> + IsType<<Self as frame_system::Config>::Event>;
```
Comment (thiolliere) [inline frame/assets/src/lib.rs:152]: IsType already bounds Into
```suggestion
		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
```

**Review by thiolliere — APPROVED** [review]

**Comment by dvdplm** [comment]
> But at the same time this is a breaking RawEvent as its generic are different

Right, so people would get a compiler error and also the deprecation message? That seems good.

**Review by ascjones — APPROVED** [review]
LGTM :rocket:

**Comment by thiolliere** [comment]
bot merge

**Comment by dvdplm** [comment]
@thiolliere I added the `D1-runtime-migration` label here but I don't think that is actually needed for `pallet-assets` (not used in Polkadot). Do you agree?

**Comment by thiolliere** [comment]
> @thiolliere I added the `D1-runtime-migration` label here but I don't think that is actually needed for `pallet-assets` (not used in Polkadot). Do you agree?

Yes I think it is better to have all the commit with potential need for migration labelled as such. It is useful for other builders

**Pull request merged**